File: swift/QGLeetCode-Swift/QGLeetCode-Swift/LC669.swift
//
//  LC669.swift
//  QGLeetCode-Swift
//
//  Created by 宇园 on 2022/9/10.
//

import Foundation

class LC669 {
    /// 669. 修剪二叉搜索树
    /**
     * Definition for a binary tree node.
     */
    public class TreeNode {
        public var val: Int
        public var left: TreeNode?
        public var right: TreeNode?
        public init() { self.val = 0; self.left = nil; self.right = nil; }
        public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
        public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
            self.val = val
            self.left = left
            self.right = right
        }
    }
    
    class Solution {
        func trimBST(_ root: TreeNode?, _ low: Int, _ high: Int) -> TreeNode? {
            let prehead = TreeNode(Int.max, root, nil)
            
            func trim(_ node: TreeNode?, _ parent: TreeNode) {
                if node == nil { return }
                
                let node = node!
                if low <= node.val && node.val <= high {
                    trim(node.left, node)
                    trim(node.right, node)
                } else if node.val < low {
                    if node === parent.left {
                        parent.left = node.right
                    } else {
                        parent.right = node.right
                    }
                    
                    trim(node.right, parent)
                } else { // node.val > high
                    if node === parent.left {
                        parent.left = node.left
                    } else {
                        parent.right = node.left
                    }
                    
                    trim(node.left, parent)
                }
            }
            
            trim(root, prehead)
            return prehead.left
        }
    }
}

File: src/main/kotlin/LC669.kt
class LC669 {
    /**
     * Example:
     * var ti = TreeNode(5)
     * var v = ti.`val`
     * Definition for a binary tree node.
     */
      class TreeNode(var `val`: Int) {
          var left: TreeNode? = null
          var right: TreeNode? = null
      }

    class Solution {
        fun trimBST(root: TreeNode?, low: Int, high: Int): TreeNode? {
            val prehead = TreeNode(Int.MAX_VALUE)
            prehead.left = root

            fun trim(node: TreeNode?, parent: TreeNode) {
                if (node == null) return

                if (low <= node.`val` && node.`val` <= high) {
                    trim(node.left, node)
                    trim(node.right, node)
                } else if (node.`val` < low) {
                    if (parent.left == node) {
                        parent.left = node.right
                    } else {
                        parent.right = node.right
                    }
                    trim(node.right, parent)
                } else {
                    if (parent.left == node) {
                        parent.left = node.left
                    } else {
                        parent.right = node.left
                    }
                    trim(node.left, parent)
                }
            }

            trim(root, prehead)
            return prehead.left
        }
    }
}